org.eclipse.xtext.testlanguages.ide/src/org/eclipse/xtext/testlanguages/fileAware/ide/refactoring/FileAwareTestLanguageResourceRelocationStrategy.xtend

```diff
@@ -0,0 +1,29 @@
+package org.eclipse.xtext.testlanguages.fileAware.ide.refactoring
+
+import com.google.inject.Inject
+import org.eclipse.xtext.ide.refactoring.IResourceRelocationStrategy
+import org.eclipse.xtext.ide.refactoring.ResourceRelocationChange
+import org.eclipse.xtext.ide.refactoring.ResourceRelocationContext
+import org.eclipse.xtext.resource.IResourceServiceProvider
+import org.eclipse.xtext.testlanguages.fileAware.fileAware.PackageDeclaration
+
+class FileAwareTestLanguageResourceRelocationStrategy implements IResourceRelocationStrategy {
+
+	@Inject IResourceServiceProvider resourceServiceProvider
+		
+	def boolean canHandle(ResourceRelocationChange change) {
+		resourceServiceProvider.canHandle(change.fromURI)
+	}
+		
+	override applyChange(ResourceRelocationContext context) {
+		context.changes.filter[ canHandle ].forEach [ change | 
+			context.addModification(change) [ resource |
+				val rootElement = resource.contents.head
+				if (rootElement instanceof PackageDeclaration) {
+					val newPackage = change.toURI.trimSegments(1).segmentsList.drop(2).join('.')
+					rootElement.name = newPackage
+				}
+			]			
+		]
```

Comment:
beautiful!

I'm undecided if we should support the use case to load a resource without recording a modification on it. 
Example: Load it, inspect model, discover nothing needs to be renamed, do nothing. 
Creating an empty recording will never be as fast as not creating a recording. Not clear though how much the difference is.

Reply:
We could make it protected for now.